File: EntitySystem2/ScoreManager.cpp
#include "ScoreManager.h"

void ScoreManager::OnCollision(EntityW::EventSp e) {
	auto collisionEvent = std::dynamic_pointer_cast<CollisionEvent>(e);
	
	EntityW::EntitySp scoringSurface = nullptr;

	if (collisionEvent->entity1->has<ScoringSurfaceComponent>()) {
		scoringSurface = collisionEvent->entity1;
	}
	else if (collisionEvent->entity2->has<ScoringSurfaceComponent>()) {
		scoringSurface = collisionEvent->entity2;
	}

	if (scoringSurface != nullptr) {
		auto scoring = scoringSurface->get<ScoringSurfaceComponent>();
		
		score[scoring->playerId]++;
		scoreText->get<TextComponent>()->text = std::to_string(score[0]) + " - " + std::to_string(score[1]);
		auto paddleTransform = scoring->paddle->get<TransformComponent>();
		Vector2 relativePosition = Vector2(glm::normalize(paddleTransform->position).x * (-1), .75);
		ball->attach<AttachComponent>(paddleTransform, relativePosition);
	}
}

File: EntitySystem2/scripts/scoreManager.lua
local scoreManager = {
	score = { 0, 0 },

	--Collision handler for scoring surfaces
	--Adds point for player and attach ball to the other
	onScoreCollision = function(self, event)
		local scoringSurface

		if event.entity1:has(Components.ScoringSurface) then
			scoringSurface = event.entity1
		elseif event.entity2:has(Components.ScoringSurface) then
			scoringSurface = event.entity2
		end

		if scoringSurface ~= nil then
			local scoring = scoringSurface:get(Components.ScoringSurface)
			self.score[scoring.playerId + 1] = self.score[scoring.playerId + 1] + 1
			self.scoreText:get(Components.Text).text = self.score[1] .. " - " .. self.score[2]

			local paddleTransform = scoring.paddle:get(Components.Transform)

			self.ball:attach(Components.Attach, {
				parentTransform = paddleTransform,
				relativePosition = { x = glm.normalize(paddleTransform.position).x * (-1), y = 0.75 },
				paddle = scoring.paddle
			})

			local ballVelocity = self.ball:get(Components.Velocity)
			print(ballVelocity.velocity.x .. ' ' .. ballVelocity.velocity.y)
			ballVelocity.velocity.x = scoring.launchVelocity;
			ballVelocity.velocity.y = 0;

			if self.score[1] >= 3 or self.score[2] >= 3 then
				emit(Events.GameOver, { score = self.score })
			end
		end
	end,

	init = function(self, scoreText, ball)
		self.scoreText = scoreText
		self.ball = ball
		subscribeForObject(Events.Collision, self.onScoreCollision, self)
	end
}

return scoreManager